pkg/loadbalancer/reconciler/termination.go

```diff
@@ -0,0 +1,252 @@
+	// directly and provide custom implementations of these.
+	cell.ProvidePrivate(
+		func() netnsOps {
+			return netnsOps{
+				current: netns.Current,
+				do:      (*netns.NetNS).Do,
+				all:     netns.All,
+			}
+		},
+		func(log *slog.Logger) sockets.SocketDestroyer {
+			return socketDestroyer{log}
+		},
+	),
+
+	cell.Invoke(registerSocketTermination),
+)
+
+type socketTerminationParams struct {
+	cell.In
+
+	JobGroup  job.Group
+	DB        *statedb.DB
+	Backends  statedb.Table[*lb.Backend]
+	Log       *slog.Logger
+	Config    lb.Config
+	ExtConfig lb.ExternalConfig
+	LBMaps    maps.LBMaps
+
+	SocketDestroyer sockets.SocketDestroyer
+	NetNSOps        netnsOps
+
+	// TestSyncChan is used by the tests to synchronize with the job so it
+	// knows when to delete the backend.
+	TestSyncChan testSyncChan   `optional:"true"`
+	TestConfig   *lb.TestConfig `optional:"true"`
+}
+
+type testSyncChan chan struct{}
+
+// netnsOps captures the operations performed with network namespaces.
+// This allows the tests to inject their own implementation.
+type netnsOps struct {
+	current func() (*netns.NetNS, error)
+	do      func(*netns.NetNS, func() error) error
+	all     func() (iter.Seq2[string, *netns.NetNS], <-chan error)
+}
+
+type socketDestroyer struct {
+	log *slog.Logger
+}
+
+func (sd socketDestroyer) Destroy(filter sockets.SocketFilter) error {
+	return sockets.Destroy(sd.log, filter)
+}
+
+func registerSocketTermination(p socketTerminationParams) error {
+	if !p.Config.EnableExperimentalLB {
+		return nil
+	}
+
+	if p.SocketDestroyer == nil {
+		// To make the load-balancer cell easier to use in tests we don't require that
+		// SocketDestroyer is always provided.
+		if p.TestConfig == nil {
+			return fmt.Errorf("SocketDestroyer not provided and not running in tests")
+		}
+		return nil
+	}
+
+	if !(p.ExtConfig.EnableSocketLB || p.ExtConfig.BPFSocketLBHostnsOnly) {
+		return nil
+	}
+
+	p.JobGroup.Add(
+		job.OneShot(
+			"socket-termination",
+			func(ctx context.Context, h cell.Health) error {
+				return socketTerminationLoop(p, ctx, h)
+			},
+		))
+
+	return nil
+}
+
+func socketTerminationLoop(p socketTerminationParams, ctx context.Context, health cell.Health) error {
+	wtxn := p.DB.WriteTxn(p.Backends)
+	changeIter, err := p.Backends.Changes(wtxn)
+	wtxn.Commit()
+	if err != nil {
+		return err
+	}
+
+	// Process the changes in batches every 50 milliseconds.
+	limiter := rate.NewLimiter(50*time.Millisecond, 1)
+	defer limiter.Stop()
+
+	for {
+		changes, watch := changeIter.Next(p.DB.ReadTxn())
+		for change := range changes {
+			backend := change.Object
+
+			if p.TestSyncChan != nil {
+				close(p.TestSyncChan)
+				p.TestSyncChan = nil
+			}
+
+			// Terminate the sockets connected to backends that have been either
+			// deleted or which are no longer considered viable.
+			if change.Deleted || !backend.IsAlive() {
+				opSupported := terminateUDPConnectionsToBackend(p, backend.Address)
```

Comment:
Would it be better to just probe for this kernel functionality once and then avoid retrying this and then avoid having these error logs periodically appear?

Reply:
This was how the original implemented it and didn't feel like diving into how to do this differently when porting this. This is not retrying as it's exiting the one-off job here.

Reply:
Yeah that makes sense to me. Would it be OK if you'd do that change in your PR once this is in?

Reply:
Yeah, sure. I will likely drop the `struct socketDestroyer` as when I rebase on this, since SocketDestroyer is `Provide`ed elsewhere. We can discuss more after merging this.